// @emailjs/nodejs/mjs/api/send_json.js
import { request } from 'https';
import { store } from '../store/store.js';
import { EmailJSResponseStatus } from '../models/emailjs_response_status.js';
export const sendJSON = (params) => {
    const options = {
        host: store._host,
        path: '/api/v1.0/email/send',
        port: 443,
        method: 'POST',
        headers: {
            'Content-Type': 'application/json',
        },
    };
    return new Promise((resolve, reject) => {
        const req = request(options, (res) => {
            const chunks = [];
            res.on('data', (chunk) => chunks.push(chunk));
            res.on('end', () => {
                const data = Buffer.concat(chunks).toString();
                if (res.statusCode === 200) {
                    resolve(new EmailJSResponseStatus(res.statusCode, data));
                }
                else {
                    reject(new EmailJSResponseStatus(res.statusCode, data));
                }
            });
            res.on('error', (error) => {
                reject(error);
            });
        });
        req.on('error', (error) => {
            reject(error);
        });
        if (params) {
            req.write(params);
        }
        req.end();
    });
};
