// @emailjs/nodejs/cjs/methods/send/send.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.send = void 0;
const store_js_1 = require("../../store/store.js");
const validate_params_js_1 = require("../../utils/validate_params.js");
const send_json_js_1 = require("../../api/send_json.js");
/**
 * Send a template to the specific EmailJS service
 * @param {string} serviceID - the EmailJS service ID
 * @param {string} templateID - the EmailJS template ID
 * @param {object} templatePrams - the template params, what will be set to the EmailJS template
 * @param {Options} options - the EmailJS options
 * @returns {Promise<string>}
 */
const send = (serviceID, templateID, templatePrams, options) => {
    const pubKey = (options === null || options === void 0 ? void 0 : options.publicKey) || store_js_1.store._publicKey;
    const prKey = (options === null || options === void 0 ? void 0 : options.privateKey) || store_js_1.store._privateKey;
    (0, validate_params_js_1.validateParams)(pubKey, serviceID, templateID);
    const params = {
        lib_version: '2.2.0',
        service_id: serviceID,
        template_id: templateID,
        user_id: pubKey,
        accessToken: prKey,
        template_params: templatePrams,
    };
    return (0, send_json_js_1.sendJSON)(JSON.stringify(params));
};
exports.send = send;
