// @emailjs/nodejs/cjs/api/send_json.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.sendJSON = void 0;
const https_1 = require("https");
const store_js_1 = require("../store/store.js");
const emailjs_response_status_js_1 = require("../models/emailjs_response_status.js");
const sendJSON = (params) => {
    const options = {
        host: store_js_1.store._host,
        path: '/api/v1.0/email/send',
        port: 443,
        method: 'POST',
        headers: {
            'Content-Type': 'application/json',
        },
    };
    return new Promise((resolve, reject) => {
        const req = (0, https_1.request)(options, (res) => {
            const chunks = [];
            res.on('data', (chunk) => chunks.push(chunk));
            res.on('end', () => {
                const data = Buffer.concat(chunks).toString();
                if (res.statusCode === 200) {
                    resolve(new emailjs_response_status_js_1.EmailJSResponseStatus(res.statusCode, data));
                }
                else {
                    reject(new emailjs_response_status_js_1.EmailJSResponseStatus(res.statusCode, data));
                }
            });
            res.on('error', (error) => {
                reject(error);
            });
        });
        req.on('error', (error) => {
            reject(error);
        });
        if (params) {
            req.write(params);
        }
        req.end();
    });
};
exports.sendJSON = sendJSON;
